File: 215. Kth Largest Element in an Array/solution.py
import heapq
from typing import List


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        """
        time: O(n log k)
        space: O(k)
        """
        heap = []
        count = 0

        for num in nums:
            heapq.heappush(heap, num)
            count += 1
            if count > k:
                heapq.heappop(heap)
                count -= 1
        return heap[0]

File: 215. Kth Largest Element in an Array/solution.go
package main

import "container/heap"

func findKthLargest(nums []int, k int) int {
    h := IntHeap(nums[:k])
	heap.Init(&h)

	for _, num := range nums {
		if num > h[0] {
			heap.Push(&h, num)
			heap.Pop(&h)
		}
	}
	return h[0]
}

type IntHeap []int

func (h IntHeap) Len() int { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
    *h = append(*h, x.(int))
}

func (h *IntHeap) Pop() interface{} {
    n := len(*h)
	x := (*h)[n-1]
	*h = (*h)[0:n-1]
	return x
}